NitroxServer/GameLogic/ScheduleKeeper.cs

```diff
@@ -0,0 +1,209 @@
+                
+                Schedule.Add(scheduledGoal.GoalKey, scheduledGoal);
+            }
+        }
+
+        public List<FakeScheduledGoal> GetScheduledGoals()
+        {
+            CleanScheduledGoals();
+            return new(Schedule.Values);
+        }
+
+        public bool ContainsScheduledGoal(string goalKey)
+        {
+            return Schedule.TryGetValue(goalKey, out FakeScheduledGoal value);
+        }
+
+        public void ScheduleGoal(FakeScheduledGoal scheduledGoal)
+        {
+            // Some weird goal that activates when the player swims, tho it should not get here
+            if (scheduledGoal.GoalKey == "PlayerDiving")
+            {
+                Log.WarnOnce("Some weird thing happened: server received a goal named \"PlayerDiving\", even though the client is not supposed to send them. Maybe one of them is playing with an outdated version ?");
+                return;
+            }
+
+            // Only add if it's not in already
+            if (!Schedule.TryGetValue(scheduledGoal.GoalKey, out FakeScheduledGoal value))
+            {
+                // If it's not already in any PDA stuff (completed goals or PDALog)
+                if (!IsAlreadyRegistered(scheduledGoal.GoalKey))
+                {
+                    if (scheduledGoal.TimeExecute > CurrentTime)
+                    {
+                        Schedule.Add(scheduledGoal.GoalKey, scheduledGoal);
+                        scheduledGoals.Add(scheduledGoal);
+                    }
+                }
+            }
+        }
+
+        public void UnScheduleGoal(string goalKey)
+        {
+            UnScheduleGoal(goalKey, false);
+        }
+
+        public void UnScheduleGoal(string goalKey, bool becauseOfTime)
+        {
+            if (!Schedule.TryGetValue(goalKey, out FakeScheduledGoal scheduledGoal))
+            {
+                return;
+            }
+            // The best solution, to ensure any bad simulation of client side, is to postpone the execution
+            // If the goal is already done, no need to check anything
+            if (becauseOfTime && !IsAlreadyRegistered(goalKey))
+            {
+                scheduledGoal.TimeExecute = CurrentTime + 15;
+                if (playerManager.GetConnectedPlayers().Count > 0)
+                {
+                    playerManager.SendPacketToAllPlayers(new Schedule(CurrentTime + 15, goalKey, scheduledGoal.GoalType));
+                    // Log.Debug($"Players connected, sending them the scheduledGoal [{scheduledGoal.GoalKey}]");
+                }
+                /* else
+                {
+                    Log.Debug($"No player connected, scheduledGoal [{scheduledGoal.GoalKey}] will be postponed");
+                }*/
+                return;
+            }
+            Schedule.Remove(goalKey);
+            scheduledGoals.Remove(scheduledGoal);
+        }
+
+        public void Init(EventTriggerer eventTriggerer, PlayerManager playerManager)
```

Comment:
The Init method is called right after the PlayerManager instance is created, also it's because we can't pass the PlayerManager in the constructor of ScheduleKeeper because they both already are in the constructor of the World

Reply:
Ah ok my bad I thought `ScheduleKeeper` is initialized with AutoFac. Can't you combine the constructor and the Init() if you move ` ScheduleKeeper = new ScheduleKeeper([...}` outside of `World world = new() [...]`?